File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.java
import java.util.Scanner;
 
class Solution
{
    public static void main(String args[]) throws Exception
    {
        Scanner sc = new Scanner(System.in);
        int T;
        T=sc.nextInt();
 
for (int t = 1; t <= T; t++) {
            int N = sc.nextInt();
            int[][] snail = new int[N][N];
 
            int x = 0; int y = 0; int d =0;
            int[] dx = {1, 0, -1, 0};
            int[] dy = {0, 1, 0, -1};
 
            for (int num = 1; num < N * N + 1; num++) {
                snail[y][x] = num;
                int nx = x + dx[d];
                int ny = y + dy[d];
 
                if ((0 <= nx && nx < N && 0 <= ny && ny < N) && snail[ny][nx] == 0) {
                    x = nx;
                    y = ny;
                } else {
                    d = (d + 1) % 4;
                    x += dx[d];
                    y += dy[d];
                }
            }
 
            System.out.println("#" + t );
            for (int[] i : snail) {
                for (int j : i) {
                    System.out.print(j + " ");
                }
                System.out.println();
            }
        }
    }
}

File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.py
T = int(input())

for test_case in range(1, T + 1):
    N = int(input())
    snail = [[0] * N for _ in range(N)]

    # x, y, d(direction)
    x, y, d = 0, 0, 0

    # 우하좌상
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]

    for num in range(1, N * N + 1):
        snail[y][x] = num
        nx = x + dx[d]
        ny = y + dy[d]

        if 0 <= nx < N and 0 <= ny < N and snail[ny][nx] == 0:
            x, y = nx, ny
        else:
            d = (d + 1) % 4
            x += dx[d]
            y += dy[d]

    print(f'#{test_case}')
    for i in range(N):
        print(*snail[i])
